pkg/distributor/distributor.go

```diff
@@ -43,6 +43,7 @@ import (
 	ingester_client "github.com/grafana/mimir/pkg/ingester/client"
 	"github.com/grafana/mimir/pkg/mimirpb"
 	"github.com/grafana/mimir/pkg/util"
+	distributor_error "github.com/grafana/mimir/pkg/util/error"
```

Comment:
Several nits. They don't necessarily all apply to the import line, but it's a good place to discuss it IMO:

- Packages shouldn't steal good variable names from the code, same applies to stealing a name as the well known interface `error`.
- If you know that all usages of a package will have to use an alias, it's better to just rename the package to that name, so all usages will be consistent and easier to identify (instead of each usage having a different alias for it)
- Package names (and aliases) should be all lowercase without underscores (`distributorerror`)
- Why distributor errors are under the `util` package? They should be under `distributor`.

So, suggestion:
- Rename `package error` to `package distributorerror` (to comply with golang style guide)
- Move it under `pkg/distributor/distributorerror` (because they belong to distributor, but we don't want to create cyclic dependencies by putting them in `pkg/distributor`)
- Rename `type DistributorError` to `type Error`, so the full name would be `distributorerror.Error` (again, following the golang style guide, to avoid stuttering, as the full name of an entity includes the package name)

Reply:
I have implemented these suggestions. Thank you.

Reply:
I agree with Oleg that these errors should be under distributor package. I would go as far as move them directly into `distributor` package.

To fix the cyclic dependency, I think we should remove dependency from distributor to push package. From what I can see, we only use `push.Func` type. We could easily copy that type into `distributor` package.

Dependency from `push` to `distributor` imho makes sense, since push needs to call distributor or deal with errors. Dependency from `distributor` to `push` should be removed.

Reply:
>To fix the cyclic dependency, I think we should remove dependency from distributor to push package. From what I can see, we only use push.Func type. We could easily copy that type into distributor package.

We use `push.Request` too. It is used in `ingester.go`, `ingester_activity.go`, `api.go`. We would need to move that one too if we decide to get rid of the `push` package in `distributor`.
What would be a good place to move `push.Request` and `push.Func`? Somewhere in `util`?

Reply:
> We use `push.Request` too. It is used in `ingester.go`, `ingester_activity.go`, `api.go`. We would need to move that one too if we decide to get rid of the `push` package in `distributor`. What would be a good place to move `push.Request` and `push.Func`? Somewhere in `util`?

For `push.Request` we could define interface with 3 public methods.

> Somewhere in `util`?

`util/push` seems like a good place for sharing. (ie. where it lives now). Perhaps we could instead move the handler out of `util/push`.

In fact moving push handler under `pkg/distributor` would not be a terrible idea either, I think. The only reason why push handler was in different package is that it was reused between Cortex and Loki. But we don't need to care about that anymore.

Reply:
@pstibrany Let's do this additional refactoring in a follow-up PR. Is it OK for you?

Reply:
I'd suggest trying to do it in preliminary PR, to make changes in this PR cleaner.

Reply:
I'm not against, but I also don't see the need of moving the errors directly to the `distributor` package right now. If that causes minimal friction, we can keep the `distributorerror` package.

Reply:
@pstibrany @colega I would really prefer to do it later. First of all because it might require a big refactoring of the current PR.

Reply:
>In fact moving push handler under pkg/distributor would not be a terrible idea either, I think. The only reason why push handler was in different package is that it was reused between Cortex and Loki. But we don't need to care about that anymore.

And what about the otlp handler? Should it also be moved under `pkg/distributor` package?

Reply:
I think we can refactor moving the push handler later. There's no need to do that now. Especially since ingester also uses the push handler.

Reply:
That's interesting point. Ingester's version of push handler will never need to deal with Distributor errors ;-)

I wonder if we *need* to have ingesters using push handler in the first place. 🤔

Reply:
> And what about the otlp handler? Should it also be moved under pkg/distributor package?

I think so. But I can see an argument for `pkg/distributor/handlers` or `pkg/distributor/push` and `pkg/distributor/otlp`.

Perhaps `util/push` could have generic version of push handler, and we can inject error-mapping functions (defined in `distributor` and `ingester` packages) to it? Then distributor and ingester errors don't even need to be public. 🤔

Reply:
@pstibrany There is this usage for "testing and debugging": https://github.com/grafana/mimir/blob/main/pkg/api/api.go#L272